### PR TITLE
Request evaluator: Cut down on the cost of AnyRequest in two ways

### PR DESCRIPTION
**Pull request by jrose-apple**
- Use a non-thread-safe llvm::IntrusiveRefCntPtr instead of std::shared_ptr.

- Use DenseMap::find_as when we just need to look up a Request and not store it, to avoid copying the Request into an AnyRequest heap box.

This shaves about a second off of the 11s of time spent type-checking the standard library.

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test compiler performance

**Comment by jrose-apple** [comment]
Fix coming tomorrow—I had it ready and then my machine-at-work crashed.

**Comment by jrose-apple** [comment]
@swift-ci Please test

File: include/swift/AST/Evaluator.h [hunk 1]
```diff
@@ -283,7 +283,7 @@ class Evaluator {
   typename Request::OutputType getResultUncached(const Request &request) {
     // Clear out the dependencies on this request; we're going to recompute
     // them now anyway.
-    dependencies[request].clear();
+    dependencies[AnyRequest(request)].clear();
```
Comment (DougGregor) [inline include/swift/AST/Evaluator.h]: Is there a `find_as`-like trick we should do here?

File: include/swift/AST/Evaluator.h [hunk 2]
```diff
@@ -349,8 +347,14 @@ class Evaluator {
                          bool lastChild) const;
 
   /// Print the dependencies of the given request as a tree.
-  void printDependencies(const AnyRequest &request,
-                         llvm::raw_ostream &out) const;
+  template <typename Request>
+  void printDependencies(const Request &request, llvm::raw_ostream &out) const {
```
Comment (DougGregor) [inline include/swift/AST/Evaluator.h]: Why bother pushing this up into the header? It's not in the critical path at all.

**Review by DougGregor — APPROVED** [review]
LGTM. I didn't know about `find_as`!

File: include/swift/AST/Evaluator.h (hunk 1 above)
Comment (jrose-apple) [reply include/swift/AST/Evaluator.h]: I tried one, but it turns out this is doing double-duty in clearing existing dependencies and in inserting an entry into the table, so I left it doing the allocation. Arguably we could split that out into two operations, but I don't think we throw away cached results that often anyway.

File: include/swift/AST/Evaluator.h (hunk 2 above)
Comment (jrose-apple) [reply include/swift/AST/Evaluator.h]: I wanted to leave the AnyRequest constructor `explicit`, but everywhere that `printDependencies` is used today it takes a specific request. Other options include:

- write this as a wrapper around the original two-argument `printDependencies`
- just make all callers construct the AnyRequest explicitly

But I thought this one seemed simplest.

File: include/swift/AST/Evaluator.h (hunk 1 above)
Comment (DougGregor) [reply include/swift/AST/Evaluator.h]: Most requests are cached, so this probably doesn't happen too often. Worst case, if it becomes a problem later, we can extend `DenseMap` with API for this.

**Pull request merged**